internal/controller/apiextensions/usage/selector_test.go

```diff
@@ -0,0 +1,530 @@
+		client client.Client
+		u      *v1alpha1.Usage
+	}
+	type want struct {
+		u   *v1alpha1.Usage
+		err error
+	}
+	cases := map[string]struct {
+		reason string
+		args   args
+		want   want
+	}{
+		"AlreadyResolved": {
+			reason: "If selectors resolved already, we should do nothing.",
+			args: args{
+				u: &v1alpha1.Usage{
+					Spec: v1alpha1.UsageSpec{
+						Of: v1alpha1.Resource{
+							APIVersion: "v1",
+							Kind:       "SomeKind",
+							ResourceRef: &v1alpha1.ResourceRef{
+								Name: "some",
+							},
+							ResourceSelector: &v1alpha1.ResourceSelector{
+								MatchLabels: map[string]string{
+									"foo": "bar",
+								},
+							},
+						},
+						By: &v1alpha1.Resource{
+							APIVersion: "v1",
+							Kind:       "AnotherKind",
+							ResourceRef: &v1alpha1.ResourceRef{
+								Name: "another",
+							},
+							ResourceSelector: &v1alpha1.ResourceSelector{
+								MatchLabels: map[string]string{
+									"baz": "qux",
+								},
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				u: &v1alpha1.Usage{
+					Spec: v1alpha1.UsageSpec{
+						Of: v1alpha1.Resource{
+							APIVersion: "v1",
+							Kind:       "SomeKind",
+							ResourceRef: &v1alpha1.ResourceRef{
+								Name: "some",
+							},
+							ResourceSelector: &v1alpha1.ResourceSelector{
+								MatchLabels: map[string]string{
+									"foo": "bar",
+								},
+							},
+						},
+						By: &v1alpha1.Resource{
+							APIVersion: "v1",
+							Kind:       "AnotherKind",
+							ResourceRef: &v1alpha1.ResourceRef{
+								Name: "another",
+							},
+							ResourceSelector: &v1alpha1.ResourceSelector{
+								MatchLabels: map[string]string{
+									"baz": "qux",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		"AlreadyResolvedNoUsing": {
+			reason: "If selectors resolved already, we should do nothing.",
+			args: args{
+				u: &v1alpha1.Usage{
+					Spec: v1alpha1.UsageSpec{
+						Of: v1alpha1.Resource{
+							APIVersion: "v1",
+							Kind:       "SomeKind",
+							ResourceRef: &v1alpha1.ResourceRef{
+								Name: "some",
+							},
+							ResourceSelector: &v1alpha1.ResourceSelector{
+								MatchLabels: map[string]string{
+									"foo": "bar",
+								},
+							},
+						},
+					},
+				},
+			},
+			want: want{
+				u: &v1alpha1.Usage{
+					Spec: v1alpha1.UsageSpec{
+						Of: v1alpha1.Resource{
+							APIVersion: "v1",
+							Kind:       "SomeKind",
+							ResourceRef: &v1alpha1.ResourceRef{
+								Name: "some",
+							},
+							ResourceSelector: &v1alpha1.ResourceSelector{
+								MatchLabels: map[string]string{
+									"foo": "bar",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		"CannotResolveUsedListError": {
+			reason: "We should return error if we cannot list the used resources.",
+			args: args{
+				client: &test.MockClient{
+					MockList: func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+						return errBoom
```

Comment:
I agree that there should be a test case checking that labels are the expected ones, I'll add one. For GVK, I already have [a test case](https://github.com/crossplane/crossplane/pull/4444/files/388deafe9d08c476bed42e2fce95b97cb5768be2#diff-0c6a70b17a7962ca4c6155b36529c54e99b93098f9977fddd70d159170684613R427) checking that, named as `BothSelectorsResolved`.

However, I believe adding a check for every invocation of `List` would be a bit of overkill. The behavior we are testing here is whether our reconciler does the right thing when it gets an error from the list call, not whether it is passing the right things which deserves a dedicated test case as I mentioned above.

Reply:
you do not need a dedicated case, since you are mocking the interaction with the cluster, you need to assert two things:

* your code is calling client List method with the expected arguments, based on the test setup
* and your code is reacting properly on the data returend by the mock.

I see that we only do the later, but not the former, but we should - otherwise a number of bugs would remain uncovered. For example, what if selector calls `client.List` without setting kind and apiVersion? This test would pass anyway.

Reply:
> you do not need a dedicated case

I didn't want to mean I would create a dedicated test case for this. I mean, it should be "sufficient" to test this at least once, and I am doing it in one of the test cases, typically the one for a successful run.

> For example, what if selector calls client.List without setting kind and apiVersion? This test would pass anyway.

Test cases meant to validate our controller handling the error properly will pass, but _the test case I mentioned above will catch it_. This is why I am saying "it should be sufficient". 

Can you come up with a reasonable scenario where we miss a bug with the above approach?

Reply:
> Test cases meant to validate our controller handling the error properly will pass, but the test case I mentioned above will catch it. This is why I am saying "it should be sufficient".

Yes, correct, however with that we create dependencies between tests, i.e. we need to keep in mind if we remove some of them, we might loose some assertions about the component under test.

The cleanest would be that unit test is self-container, but I do not insists on that if it is see as a lot of work. Perhaps we should really think about using `gomocks`, since it simplifies exactly those kind of assertions.

Reply:
> The cleanest would be that unit test is self-container, but I do not insists on that if it is see as a lot of work.

Yes, it would be cleaner to do so without much effort/duplication. However, with the current approach it is becoming tedious with not too much added value. 

> Perhaps we should really think about using gomocks, since it simplifies exactly those kind of assertions.

I agree with this. We should at least discuss the tradeoffs and make a deliberate decision in general instead of discussing them in individual PRs. The reason why this didn't come up so far (at least I am not aware of) is we probably didn't miss some real bugs due to lack of input checks in unit tests to think it as an important thing to do.

Another point that I am just realizing is it is indeed not possible/feasible to do input checks with the current approach we are following. For example, I am trying to add some input checks to an Update call which is called multiple times, and I cannot find a clean way of doing so for different calls 🤷

Reply:
> We should at least discuss the tradeoffs and make a deliberate decision in general instead of discussing them in individual PRs.

So, please consider opening a dedicated issue if you feel that it worths the effort and there are some real values we can get out of it.